rd-net/Lifetimes/Lifetimes/LifetimedList.cs

```diff
@@ -57,61 +77,218 @@ public bool MoveNext()
+        }
+        finally
+        {
+          ReleaseLocalLock();
+        }
+      }
+      finally
+      {
+        ReleaseGlobalLock();
+      }
+
+      if (shouldClear)
+      {
+        // if ClearValuesIfNotAlive was called at the same time with EnsureCapacityNoLock, we may not have cleared some elements
+        // 1. Thread 1 - create a new array (items)
+        // 2. Thread 1 - copy item[0] to the new array
+        // 3. Thread 2 - terminate item[0].Lifetime
+        // 4. Thread 2 - ClearValuesIfNotAlive
+        // 5. Thread 1 - myItems = items (myItems still contains uncleared item)
+        // Memory leak
+        ClearValuesIfNotAlive();
+      }
     }        
 
     public void AddPriorityItem(ValueLifetimed<T> item)
     {
-      EnsureCapacity();
-      Array.Copy(myItems, Marker, myItems, Marker+1, mySize-Marker);
-      myItems[Marker++] = item;
-      mySize++;
+      EnterGlobalLock();
+      try
+      {
+        // AddPriorityItem is not a frequent operation, we can afford to create a new array each time
+        EnsureCapacityNoLock(true, out var items, out var marker, out var size);
+        items[marker] = item;
+        
+        EnterLocalLock();
+        try
+        {
+          // increase mySize only after inserting an element
+          mySize = size + 1;
+          myItems = items;
+          // no need to use InterlockedUpdate because we under lock
+          myState = MarkerMutexSlice.Updated(myState, marker + 1);
+        }
+        finally
+        {
+          ReleaseLocalLock();
+        }
+      }
+      finally
+      {
+        ReleaseGlobalLock();
+      }
+      
+      // AddPriorityItem always creates a new array
+      // if ClearValuesIfNotAlive was called at the same time with EnsureCapacityNoLock, we may not have cleared some elements
+      // 1. Thread 1 - create a new array (items)
+      // 2. Thread 1 - copy item[0] to the new array
+      // 3. Thread 2 - terminate item[0].Lifetime
+      // 4. Thread 2 - ClearValuesIfNotAlive
+      // 5. Thread 1 - myItems = items (myItems still contains uncleared item)
+      // Memory leak
+      ClearValuesIfNotAlive();
     }
 
-    private void EnsureCapacity()
+    private bool EnsureCapacityNoLock(bool priority, out ValueLifetimed<T>[] items, out int marker, out int size)
     {
-      if (myItems == null)
-        myItems = new ValueLifetimed<T>[1];
-      
-      if (mySize < myItems.Length)
-        return;
+      myItems ??= new ValueLifetimed<T>[1];
+
+      marker = MarkerMutexSlice[myState];
       
+      if (mySize < myItems.Length && !priority)
+      {
+        items = myItems;
+        size = mySize;
+        return false;
+      }
+     
+      var newSize = 0;
+      for (var i = 0; i < mySize; i++)
+      {
+        if (myItems[i].Lifetime.IsAlive) 
+          newSize++;
+      }
+
+      if (newSize == 0) newSize = 1;
+      else if (priority) newSize++; // for new priority item
```

Comment:
This will mean we reserve two items in the resulting array for a priority item. Is that right?

Reply:
I think that's ok, but maybe it should rewrite like this:
```
if (newSize == 0) newSize = 1;
else newSize *= 2;
...
var newItems = new ValueLifetimed<T>[newSize];      
```